index.bs

```diff
@@ -7,14 +7,160 @@ Group: browser-testing-tools
+
+A WebDriver [=session=] has a <dfn>WebSocket connection</dfn>, which
+is a network connection, and a <dfn>WebSocket URL</dfn>, which is a
+string. These are both initially null.
+
+When data is recieved on [=WebSocket connection=]
+<var>connection</var>, it must be processed according to the
+requirements of the [[!RFC6455|WebSockets]] specification. When [=a
+WebSocket message has been received=], an implementation must [=handle
+an incoming message=] with argument <var>connection</var>. When [=the
+WebSocket closing handshake is started=] an implementation must
+[=close the WebSocket connection=] with argument
+<var>connection</var>.
+
+<div algorithm>
+To <dfn>start listening for a WebSocket connection</dfn>, with
+parameter |session id|, run these steps:
+
+ 1. Set up a network connection that listens on an implementation-defined
+    hostname |host| and port |port|. The connection may TLS encrypted, in which
+    case let |secure| be <code>true</code>, otherwise let |secure| be
+    <code>false</code>. Set the [=connection=] to this network connection.
+
+ 2. Let |path| be the result of concatenating the string "<code>session/</code>"
+    with |session id|.
+
+ 3. Let |url| be the result of constructing a [=WebSocket URI=] with scheme
+    "<code>wss</code>" if |secure| is <code>true</code> or "<code>ws</code>"
+    otherwise, host |host|, port |port|, path |path| and empty query.
+
+ 4. Set the [=websocket url=] to |url|.
+
+ 5. Return |url|.
+
+</div>
+
+Note: For an [=endpoint node=] the hostname in the above steps will be
+typically be "<code>localhost</code>".
+
+<div algorithm> To <dfn>handle an incoming message</dfn> with parameter
+<var ignore>connection</var>, run these steps:
+
+Issue: Figure this out.
+
+</div>
+
+<div algorithm> To <dfn>close the WebSocket connection</dfn> with
+parameter <var>connection</var>, run these steps:
+
+ 1. Close the underlying network connection associated with
+    <var>connection</var>.
+
+ 2. If there is a WebDriver [=session=] with <var>connection</var> as
+    its [=connection=], set the [=connection=] on that [=session=] to
+    null.
+
+Issue: This should also reset any internal state
+
+</div>
+
+Note: This does not end any [=session=].
+
+Establishing a Connection {#estabishing}
+========================================
+
+WebDriver clients opt in to a bidirectional connection by requesting a
+capability with the name "<code>bidi</code>" and value
+<code>true</code>.
+
+This specification defines an
+[=additional webdriver capability=] with the [=capability name=] "<code>bidi</code>".
+
+<div algorithm="bidi capability deserialization algorithm">
+The [=additional capability deserialization algorithm=] for the
+"<code>bidi</code>" capability, with parameter |value| is:
+
+ 1. If |value| is not a boolean, return [=error=] with [=error code|code=]
+    [=invalid argument=].
+
+ 2. Return [=success=] with data |value|.
+
+</div>
+
+<div algorithm="bidi capability serialization algorithm">
```

Comment:
Just to check my understanding of how things are structured: this hook exists only to propagate the capability to *matched capabilities* in https://w3c.github.io/webdriver/#dfn-matching-capabilities. One *could* serialize the actual URL here, but it's possible this won't be in the finally chosen *matched capabilities* in https://w3c.github.io/webdriver/#dfn-capabilities-processing, so we shouldn't start listening until our "bidi new session algorithm" hook.

I think I found a bug while staring at this: https://github.com/w3c/webdriver/issues/1530

Reply:
Well we can't serialize the actual url here if it includes the session id since that isn't known yet. But basically, yes, that's correct.